pkg/controller/filter.go

```diff
@@ -44,13 +44,18 @@ func FilterRunRef(apiVersion, kind string) func(interface{}) bool {
 			// Ignore.
 			return false
 		}
-		if r == nil || r.Spec.Ref == nil {
+		if r == nil || (r.Spec.Ref == nil && r.Spec.Spec == nil) {
 			// These are invalid, but just in case they get
 			// created somehow, don't panic.
 			return false
 		}
-
-		return r.Spec.Ref.APIVersion == apiVersion && r.Spec.Ref.Kind == v1alpha1.TaskKind(kind)
+		result := false
+		if r.Spec.Ref != nil {
+			result = r.Spec.Ref.APIVersion == apiVersion && r.Spec.Ref.Kind == v1alpha1.TaskKind(kind)
+		} else if r.Spec.Spec != nil {
+			result = r.Spec.Spec.APIVersion == apiVersion && r.Spec.Spec.Kind == kind
```

Comment:
Nothing wrong with your changes here, just an observation:

`r.Spec.Ref.Kind` is of type `TaskKind` and that's why needs the conversion v/s `r.Spec.Spec.Kind` is of type `string`. `TaskKind` is also originally used to signify whether a task is `task` or `clustertask` but with custom task, it can be set to anything.